arrays.qmd

```diff
@@ -5,105 +5,234 @@
-dim(R) <- N_age
-dim(n_SI) <- N_age
-dim(n_IR) <- N_age
-dim(p_SI) <- N_age
-dim(m) <- c(N_age, N_age)
-dim(s_ij) <- c(N_age, N_age)
-dim(lambda) <- N_age
+new_S[, 1] <- S[i, j] - n_SI[i, j] - n_S_vax[i, j] + n_S_vax[i, N_vax]
+new_S[, 2:N_vax] <- S[i, j] - n_SI[i, j] - n_S_vax[i, j] + n_S_vax[i, j - 1]
+```
+which we could also write as
+```r
+new_S[, ] <- S[i, j] - n_SI[i, j] - n_S_vax[i, j]
+new_S[, 1] <- new_S[i, j] + n_S_vax[i, N_vax]
+new_S[, 2:N_vax] <- new_S[i, j] + n_S_vax[i, j - 1]
+```
+or another alternative way of writing this would be to use `if else`
+```r
+new_S[, ] <- S[i, j] - n_SI[i, j] - n_S_vax[i, j] +
+    (if (j == 1) n_S_vax[i, N_vax] else n_S_vax[i, j - 1])
 ```
+Note that in odin, an `if` always requires an `else`!
 
-Relevant changes from the above:
+## Indexing
 
+We have seen in the above examples uses of 1d and 2d arrays, making use of index variables `i` and `j`. Note that while we never use these index variables on the LHS, it is implicit that on the LHS we are indexing by `i` for 1d arrays and `i` and `j` for 2d arrays! Use of these index variables on the RHS corresponds to the indexing on the LHS.
+
+Of course you may want to use higher-dimensional arrays! We currently supporting up to 8 dimensions, with index variables `i`, `j`, `k`, `l`, `i5`, `i6`, `i7` and `i8`.
+
+Be careful with array equations! It's possible that in an array equation you end up going out of bounds with an array used on the RHS. This can crash the program when running.
+
+## Declaring dimensions
+You must declare the dimensions of all arrays within your `odin` code using a `dim` equation.
+
+Dimensions can be hardcoded:
+```r
+dim(y) <- 10
 ```
-m <- parameter() # age-structured contact matrix
-s_ij[, ] <- m[i, j] * I[j]
-lambda[] <- beta * sum(s_ij[i, ])
-p_SI[] <- 1 - exp(-lambda[i] * dt)
-update(S[]) <- S[i] - n_SI[i]
-N_age <- user()
-dim(S) <- N_age
+or dimensions themselves can be parameters:
+```r
+dim(x) <- c(nr, 5)
+nr <- parameter()
+```
+You may have array parameters, these can have dimensions explicitly declared
+```r
 dim(m) <- c(N_age, N_age)
 ```
+or they can also have their dimensions detected when the parameters are passed into the system - this still needs a `dim` equation where you must explicitly state the rank:
+```r
+dim(m) <- parameter(rank = 2)
+```
+
+## Semantics of random number draws
+
+Stochastic functions are called for each element in an array they are assigned to, at each time.  So here:
+
+```r
+x[] <- Normal(0, 1)
+```
+
+`x` will be filled with each element having a different draw from a standard normal.  In contrast, in:
+
+```r
+a <- Normal(0, 1)
+x[] <- a
+```
+
+`x` will be a vector where every element is the same, the result of a *single* draw from a standard normal.
+
+
+## Summing over arrays
+Frequently, you will want to take a sum over an array, or part of an array, using `sum`.  To sum over all elements of an array, use `sum()` with the name of the array you would like to sum over:
```

Comment:
This section comes *after* the first bit of array modelling, so suggest a reordering?  Perhaps build up to the SIR model you have by writing a simple model (e.g., something like SIR by hand for two cases, manually summing for the FOI, then merging, or even simpler two logistic growth curves then introduce the idea of summing)